storage/remote/read.go

```diff
@@ -93,11 +96,29 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
 	if err != nil {
 		return nil, err
 	}
-
 	removeLabels(res, added)
+	validateNamesAndValues(res)
+
 	return res, err
 }
 
+// validateNamesAndValues rejects invalid metric/label names returned from remote read.
+func validateNamesAndValues(res model.Matrix) (err error) {
+	for i, r := range res {
+		for name, value := range r.Metric {
+			if !strutil.IsValidLabelName(string(name)) || !utf8.ValidString(string(value)) || !model.IsValidMetricName(value) {
```

Comment:
model.IsValidMetricName should only be applied to the __name__ label, not all labels.